docs/src/main/asciidoc/cache.adoc

```diff
@@ -366,6 +378,112 @@ public class CachedService {
 <3> Calling this method WILL invalidate values cached by the `load` method because the key elements order is the same.
 <4> Calling this method WILL NOT invalidate values cached by the `load` method because the key elements order is different.
 
+=== Generating a cache key with `CacheKeyGenerator`
+
+You may want to include more than the arguments of a method into a cache key.
+This can be done by implementing the `io.quarkus.cache.CacheKeyGenerator` interface and declaring that implementation in the `keyGenerator` field of a `@CacheResult` or `@CacheInvalidate` annotation.
+
+If a CDI scope is declared on a key generator, then it will be injected as a CDI bean during the cache key computation.
+Otherwise, the key generator will be instantiated using its default constructor.
+All CDI scopes supported by Quarkus can be used on a key generator.
+
+[WARNING]
+====
+All cache key generators must have a default no-arg constructor.
```

Comment:
If it's a CDI bean, why is this a requirement?

Reply:
Good question. I guess I should lift that requirement for CDI beans.

I wonder what's the best approach to do that at build time though. Looking for CDI scopes annotations on the generator class isn't enough since the generator could still be a bean produced by CDI even if it didn't have any scope declared on its class. Is there an easy way to determine from a `@BuildStep` if an instance is a bean managed by CDI?

Reply:
I think there is, but lets see what @mkouba has to say :)

Reply:
Maybe `Arc.container().instance(keyGeneratorClass).isAvailable()` works at build time, I didn't test it. I'll check that right away.

Reply:
Have you looked at `BeanDiscoveryFinishedBuildItem`?

Reply:
No, but that name sounds promising 😃

Reply:
I updated `CacheProcessor`. CDI beans are now excluded from the no-arg constructor check for key generators.